server/src/editor.js

```diff
@@ -0,0 +1,140 @@
+    this.imageContext.drawImage(highlighter, 0, 0);
+    let dataUrl = this.imageCanvas.toDataURL();
+    this.props.onClickSave(dataUrl);
+  }
+
+  onClickHighlight() {
+    if (this.state.tool != 'highlighter') {
+      this.setState({tool: 'highlighter', highlighter: 'active', pen: 'inactive'});
+      sendEvent("highlighter-select", "annotation-toolbar");
+    } else {
+      this.setState({tool: 'none', highlighter: 'inactive'});
+      sendEvent("highlighter-deselect", "annotation-toolbar");
+    }
+  }
+
+  onClickPen() {
+    if (this.state.tool != 'pen') {
+      this.setState({tool: 'pen', pen: 'active', highlighter: 'inactive'});
+      sendEvent("pen-select", "annotation-toolbar");
+    } else {
+      this.setState({tool: 'none', pen: 'inactive'});
+      sendEvent("pen-deselect", "annotation-toolbar");
+    }
+  }
+
+  componentDidMount() {
+    editor = this.refs.editor;
+    context = editor.getContext('2d');
+    highlighter = this.refs.highlighter;
+    highlightContext = highlighter.getContext('2d');
+    let imageCanvas = this.refs.image;
+    this.imageCanvas = imageCanvas
+    let imageContext = imageCanvas.getContext('2d');
+    this.imageContext = imageContext;
+    // From https://blog.headspin.com/?p=464, we oversample the canvas for improved image quality
+    let img = new Image();
+    img.src = this.props.clip.image.url;
+    let width = this.props.clip.image.dimensions.x;
+    let height = this.props.clip.image.dimensions.y;
+    this.imageContext.scale(2, 2);
+    this.imageContext.drawImage(img, 0, 0, width, height);
+  }
+
+  edit() {
+    pos = { x: 0, y: 0 };
+    if (this.state.tool == 'highlighter') {
+      highlightContext.lineWidth = 20;
+      highlightContext.strokeStyle = '#ff0';
+      drawContext = highlightContext;
+    } else if (this.state.tool == 'pen') {
+      context.strokeStyle = this.state.color;
+      highlightContext.strokeStyle = this.state.color;
+      context.globalCompositeOperation = 'source-over';
+      drawContext = context;
+    }
+    context.lineWidth = this.state.size;
+    document.querySelector("#canvas-container").addEventListener("mousemove", this.draw);
+    document.querySelector("#canvas-container").addEventListener("mousedown", this.setPosition);
+    document.querySelector("#canvas-container").addEventListener("mouseenter", this.setPosition);
+    if (this.state.tool == 'none') {
+      document.querySelector("#canvas-container").removeEventListener("mousemove", this.draw);
+      document.querySelector("#canvas-container").removeEventListener("mousedown", this.setPosition);
+      document.querySelector("#canvas-container").removeEventListener("mouseenter", this.setPosition);
+    }
+  }
+
+  setPosition(e) {
+    var rect = editor.getBoundingClientRect();
+    pos.x = e.clientX - rect.left,
+    pos.y = e.clientY - rect.top
+  }
+
+  draw(e) {
+    if (e.buttons !== 1) return;
+    drawContext.beginPath();
+
+    drawContext.lineCap = 'round';
```

Comment:
I wonder if a vertical bar would be more appropriate than a circle? Eh, probably not what people expect.